agent/lib_aws_sdk_php.c

```diff
@@ -14,10 +14,420 @@
+    'Entries' => array(
+        array(
+            // Id is required
+            'Id' => 'string',
+            // MessageBody is required
+            'MessageBody' => 'string',
+            'DelaySeconds' => integer,
+            'MessageAttributes' => array(
+                // Associative array of custom 'String' key names
+                'String' => array(
+                    'StringValue' => 'string',
+                    'BinaryValue' => 'string',
+                    'StringListValues' => array('string', ... ),
+                    'BinaryListValues' => array('string', ... ),
+                    // DataType is required
+                    'DataType' => 'string',
+                ),
+                // ... repeated
+            ),
+        ),
+        // ... repeated
+    ),
+));
+
+//clang-format on
+*/
+void nr_lib_aws_sdk_php_sqs_handle(nr_segment_t* segment,
+                                   char* command_name_string,
+                                   size_t command_name_len,
+                                   NR_EXECUTE_PROTO) {
+  char* command_arg_value = NULL;
+  nr_segment_message_params_t message_params = {
+      .library = "SQS",
+      .destination_type = NR_MESSAGE_DESTINATION_TYPE_QUEUE,
+      .messaging_system = "aws_sqs",
+  };
+  nr_segment_cloud_attrs_t cloud_attrs = {0};
+
+  if (NULL == segment) {
+    return;
+  }
+
+  if (NULL == command_name_string || 0 == command_name_len) {
+    return;
+  }
+
+  /* Determine if we instrument this command. */
+  if (command_name_len == AWS_SQS_SEND_MESSAGE_BATCH_COMMAND_LEN
+      && 0
+             == nr_strncmp(AWS_SQS_SEND_MESSAGE_BATCH_COMMAND,
+                           command_name_string,
+                           AWS_SQS_SEND_MESSAGE_BATCH_COMMAND_LEN)) {
+    message_params.message_action = NR_SPANKIND_PRODUCER;
+  } else if (command_name_len == AWS_SQS_SEND_MESSAGE_COMMAND_LEN
+             && 0
+                    == nr_strncmp(AWS_SQS_SEND_MESSAGE_COMMAND,
+                                  command_name_string,
+                                  AWS_SQS_SEND_MESSAGE_COMMAND_LEN)) {
+    message_params.message_action = NR_SPANKIND_PRODUCER;
+  } else if (command_name_len == AWS_SQS_RECEIVE_MESSAGE_COMMAND_LEN
+             && 0
+                    == nr_strncmp(AWS_SQS_RECEIVE_MESSAGE_COMMAND,
+                                  command_name_string,
+                                  AWS_SQS_RECEIVE_MESSAGE_COMMAND_LEN)) {
+    message_params.message_action = NR_SPANKIND_CONSUMER;
+  } else {
+    /* Nothing to do here so exit. */
+    return;
+  }
+
+  cloud_attrs.aws_operation = command_name_string;
+
+  command_arg_value = nr_lib_aws_sdk_php_get_command_arg_value(
+      AWS_SDK_PHP_SQSCLIENT_QUEUEURL_ARG, NR_EXECUTE_ORIG_ARGS);
+
+  /*
+   *  nr_lib_aws_sdk_php_sqs_parse_queueurl checks for NULL so safe pass
+   * command_arg_value directly in.
+   */
+  nr_lib_aws_sdk_php_sqs_parse_queueurl(command_arg_value, &message_params,
```

Comment:
Consider an optimization where a copy of `command_arg_value` is passed in and then the parsing function can modify this buffer in place and return pointers to the fields within this buffer to avoid strdup'ing the 3 fields in the parsing function.  Also removes the need to free these later.  Just need to document the parse function REQUIRES a copy that can be modified.

Reply:
Please see: [f48df3f](https://github.com/newrelic/newrelic-php-agent/pull/1003/commits/f48df3f365177c9a137f6d11f5a9261681469b80)

Reply:
Hold off on looking at this, it needs additional rework.

Reply:
Tests needed to be updated for this refactor to pass modifiable strings not literals.
[0b194c2](https://github.com/newrelic/newrelic-php-agent/pull/1003/commits/0b194c2382d09799451eee9274fe9ddd02531586)